tool/src/demo_gen/terminus.rs

```diff
@@ -0,0 +1,380 @@
+
+    /// Final result of recursively calling [`RenderTerminusContext::evaluate_constructor`] on [`MethodDependency`]
+    pub node_call_stack: String,
+
+    /// Are we a typescript file? Set by [`super::WebDemoGenerationContext::init`]
+    pub typescript: bool,
+
+    /// List of JS imports that this terminus needs.
+    pub imports: BTreeSet<String>,
+}
+
+impl<'ctx, 'tcx> RenderTerminusContext<'ctx, 'tcx> {
+    /// Create a Render Terminus .js file from a method.
+    /// We define this (for now) as any function that outputs [`hir::SuccessType::Write`]
+    pub fn evaluate(&mut self, type_name: String, method: &Method) {
+        // TODO: I think it would be nice to have a stack of the current namespace a given parameter.
+        // For instance, ICU4XFixedDecimalFormatter.formatWrite() needs a constructed ICU4XFixedDecimal, which takes an i32 called v as input.
+        // Someone just trying to read the .d.ts file will only see function formatWrite(v: number); which doesn't really help them figure out where that's from or why it's there.
+
+        // Not making this as part of the RenderTerminusContext because we want each evaluation to have a specific node,
+        // which I find easier easier to represent as a parameter to each function than something like an updating the current node in the struct.
+        let mut root =
+            MethodDependency::new(self.get_constructor_js(type_name.to_string(), method));
+
+        // And then we just treat the terminus as a regular constructor method:
+        self.terminus_info.node_call_stack = self.evaluate_constructor(method, &mut root);
+
+        let type_n = type_name.clone();
+        let format = self
+            .formatter
+            .fmt_import_statement(&type_n, false, "./js/".into());
+
+        self.terminus_info.imports.insert(format);
+    }
+
+    /// Currently unused, plan to hopefully use this in the future for quickly grabbing parameter information.
+    fn _get_type_demo_attrs(&self, ty: &Type) -> Option<DemoInfo> {
+        ty.id()
+            .map(|id| self.tcx.resolve_type(id).attrs().demo_attrs.clone())
+    }
+
+    /// Take a parameter passed to a terminus (or a constructor), and either:
+    /// 1. Add it to the list of parameters that the terminus function takes for the render engine to call.
+    /// 2. Go a step deeper and look at its possible constructors to call evaluate_param on.
+    ///
+    /// `node` - Represents the current function of the parameter we're evaluating. See [`MethodDependency`] for more on its purpose.
+    fn evaluate_param(
+        &mut self,
+        param_type: &Type,
+        param_name: String,
+        node: &mut MethodDependency,
+        method_attrs: DemoInfo,
+    ) {
+        let attrs = Some(method_attrs); /*self.get_type_demo_attrs(param_type).or(Some(method_attrs));*/
+        // Helper function for quickly passing a parameter to both our node and the render terminus.
+        let out_param = |type_name| {
```

Comment:
nit: `flush_param()`? or `append_param()`

Reply:
Also I would be fine with this being a full method that takes in `self` and `node` as arguments.